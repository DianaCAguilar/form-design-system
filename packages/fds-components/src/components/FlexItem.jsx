import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

/**
 * @description map of justify prop values to class names
 */
const JustifyPropMap = {
  flexEnd: 'flexItem--justifyEnd',
  center: 'flexItem--justifyCenter',
  spaceBetween: 'flexItem--justifySpaceBetween',
  spaceAround: 'flexItem--justifySpaceAround',
};

/**
 * @description map of align prop values to class names
 */
const AlignPropMap = {
  flexStart: 'flexItem--alignStart',
  center: 'flexItem--alignCenter',
  flexEnd: 'flexItem--alignEnd',
};

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const FlexItem = (props) => {
  const classNames = cx(
    'flexItem',
    {
      'flexItem--shrink': props.shrink,
    },
    {
      [JustifyPropMap[props.justify]]: props.justify,
      [AlignPropMap[props.align]]: props.align,
    },
    props.classNames
  );

  return <div className={classNames}>{props.children}</div>;
};

FlexItem.defaultProps = {
  shrink: false,
};

FlexItem.propTypes = {
  shrink: PropTypes.bool,
  justify: PropTypes.oneOf(Object.keys(JustifyPropMap)),
  align: PropTypes.oneOf(Object.keys(AlignPropMap)),
  classNames: PropTypes.string,
  children: PropTypes.node,
};

export default FlexItem;
