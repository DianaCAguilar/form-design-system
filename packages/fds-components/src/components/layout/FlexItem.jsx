import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

/**
 * @description map of justify prop values to class names
 */
const JustifyPropMap = {
  start: 'flexItem--justifyStart',
  end: 'flexItem--justifyEnd',
  center: 'flexItem--justifyCenter',
  spaceBetween: 'flexItem--justifySpaceBetween',
  spaceAround: 'flexItem--justifySpaceAround',
};

/**
 * @description map of align prop values to class names
 */
const AlignPropMap = {
  start: 'flexItem--alignStart',
  end: 'flexItem--alignEnd',
  center: 'flexItem--alignCenter',
};

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const FlexItem = (props) => {
  const classNames = cx(
    'flexItem',
    AlignPropMap[props.align],
    JustifyPropMap[props.justify],
    {
      'flexItem--shrink': props.shrink,
    },
    props.className
  );

  return <div className={classNames}>{props.children}</div>;
};

FlexItem.defaultProps = {
  shrink: false,
  align: 'start',
  justify: 'start',
};

FlexItem.propTypes = {
  justify: PropTypes.oneOf(Object.keys(JustifyPropMap)),
  align: PropTypes.oneOf(Object.keys(AlignPropMap)),

  /** When set, the `FlexItem` will shrink to content size */
  shrink: PropTypes.bool,

  /** Classes to pass to flex child */
  className: PropTypes.string,

  /** React children */
  children: PropTypes.node,
};

export default FlexItem;
