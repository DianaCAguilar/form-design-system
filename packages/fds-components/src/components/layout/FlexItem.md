
#### Basic example

```js
<div className="debug--classes">
  <Flex>
    <FlexItem shrink>
      <img src="https://via.placeholder.com/60/a02385/ffffff" />
    </FlexItem>
    <FlexItem>
      <p>Here is a company description, lol</p>
    </FlexItem>
    <FlexItem shrink>
      <button type="button">Edit</button>
    </FlexItem>
    <FlexItem shrink>
      <button type="button">Delete</button>
    </FlexItem>
  </Flex>
</div>
```
