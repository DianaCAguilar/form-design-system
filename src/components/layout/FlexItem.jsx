import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

/**
 * @description map of justify prop values to class names
 */
const JustifyPropMap = {
  start: 'flexItem--justifyStart',
  end: 'flexItem--justifyEnd',
  center: 'flexItem--justifyCenter',
  spaceBetween: 'flexItem--justifySpaceBetween',
  spaceAround: 'flexItem--justifySpaceAround',
};

/**
 * @description map of align prop values to class names
 */
const AlignPropMap = {
  start: 'flexItem--alignStart',
  end: 'flexItem--alignEnd',
  center: 'flexItem--alignCenter',
};

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const FlexItem = (props) => {
  const classNames = cx(
    'flexItem',
    {
      'flexItem--shrink': props.shrink,
    },
    {
      [JustifyPropMap[props.justify]]: props.justify,
      [AlignPropMap[props.align]]: props.align,
    },
    props.className
  );

  if (
    process.env.NODE_ENV === 'development' &&
    // eslint-disable-next-line react/prop-types
    (props.dataTest || props['data-test'] || props.className)
  ) {
    // eslint-disable-next-line no-console
    console.warn(
      'You can no longer put data attributes or classNames on FlexItem (Try putting these on a div inside FlexItem)'
    );
  }

  return (
    <div className={classNames} data-test={props.dataTest}>
      {props.children}
    </div>
  );
};

FlexItem.defaultProps = {
  shrink: false,
  dataTest: '',
};

FlexItem.propTypes = {
  /** Sets standard justify-content */
  justify: PropTypes.oneOf(Object.keys(JustifyPropMap)),
  /** Sets standard align-items */
  align: PropTypes.oneOf(Object.keys(AlignPropMap)),

  /** When set, the `FlexItem` will shrink to content size */
  shrink: PropTypes.bool,

  /** Classes to pass to flex child */
  className: PropTypes.string,

  /** React children */
  children: PropTypes.node,

  /** data-test to pass to flex child */
  dataTest: PropTypes.string,
};

export default FlexItem;
