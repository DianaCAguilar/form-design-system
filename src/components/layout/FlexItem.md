## Shrinking & growing a `FlexItem`
By default, all `FlexItem` components will grow to fill remaining space within the parent
`Flex`. The boolean prop `shrink` will cause a `FlexItem` to shrink to its content size.


## Controlling alignment/justification
You can use the same `align` and `justify` prop interface from `Flex` to control alignment and justification of individual `FlexItem` components.

