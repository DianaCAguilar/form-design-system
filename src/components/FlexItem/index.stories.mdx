import React from 'react';

import Flex from 'components/Flex';
import FlexItem from '.';

const Template = (args) => (
  <div style={{ width: '600px' }} className="debug--flex">
    <Flex>
      <FlexItem {...args}>
        <img src="https://place-hold.it/40x40/a02385/fff&text=IMG" />
      </FlexItem>
      <FlexItem {...args}>
        <p>Dogs</p>
      </FlexItem>
      <FlexItem {...args}>
        <p>Cats</p>
      </FlexItem>
    </Flex>
  </div>
)

export const Primary = Template.bind({});
Primary.args = {};

export default {
  component: FlexItem,
  title: 'components/FlexItem',
  parameters: {
    docs: {
      description: {
        component: 'Direct parent must be `<Flex>`. By default, all `FlexItem` components will grow to fill remaining space within the parent `Flex`.',
      },
    },
  },
};
