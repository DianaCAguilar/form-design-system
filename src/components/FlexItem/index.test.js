import React from 'react';
import { shallow } from 'enzyme';

import FlexItem from '.';

describe('Flex component', () => {
  it('matches snapshot (default props)', () => {
    expect(
      shallow(
        <FlexItem>
          <div />
        </FlexItem>
      )
    ).toMatchSnapshot();
  });

  it('matches snapshot (no child)', () => {
    expect(shallow(<FlexItem></FlexItem>)).toMatchSnapshot();
  });

  it('matches snapshot (set all props)', () => {
    expect(
      shallow(<FlexItem shrink align="center" justify="spaceBetween" />)
    ).toMatchSnapshot();
  });
});
