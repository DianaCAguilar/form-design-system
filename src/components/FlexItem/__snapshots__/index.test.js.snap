// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex component matches snapshot (default props) 1`] = `
<div
  className="flexItem"
>
  <div />
</div>
`;

exports[`Flex component matches snapshot (no child) 1`] = `""`;

exports[`Flex component matches snapshot (set all props) 1`] = `""`;
